archive/walk.go

```diff
@@ -245,6 +250,110 @@ func (c Chunk) Range() string {
 	return fmt.Sprintf("[%d-%d]", c.First, c.Last)
 }
 
+type chunkReader struct {
+	io.Reader
+	io.Closer
+	totalSize int64
+	readSize  int64
+}
+
+// newChunkReader returns an io.ReadCloser for this chunk. If the chunk has a seek
+// index and if the provided span skips part of the chunk, the seek index will
+// be used to limit the reading window of the chunk's reader.
+func newChunkReader(ctx context.Context, chunk Chunk, ark *Archive, span nano.Span) (*chunkReader, error) {
+	cspan := chunk.Span()
+	span = cspan.Intersect(span)
+	if span.Dur == 0 {
+		return nil, nil
+	}
+	uri := chunk.Path(ark)
+	stat, err := iosrc.Stat(ctx, uri)
+	if err != nil {
+		return nil, err
+	}
+	r, err := iosrc.NewReader(ctx, chunk.Path(ark))
+	if err != nil {
+		return nil, err
+	}
+	cr := &chunkReader{
+		Reader:    r,
+		Closer:    r,
+		totalSize: stat.Size(),
+		readSize:  stat.Size(),
+	}
+	if span == cspan {
+		return cr, nil
+	}
+	finder, err := microindex.NewFinder(ctx, resolver.NewContext(), chunk.seekIndexPath(ark))
+	if err != nil {
+		return nil, err
+	}
+	start, end, err := spanOffsets(ctx, span, finder)
+	if err != nil {
+		return nil, err
+	}
+	if start > 0 {
+		if _, err := r.Seek(start, io.SeekStart); err != nil {
+			return nil, err
+		}
+	}
+	if end > cr.totalSize {
+		end = cr.totalSize
+		cr.Reader = io.LimitReader(r, end-start)
```

Comment:
Shouldn't this happen outside of this `if` branch? If end is less than the size of the file, then you'd want to use a limit reader as well?

Reply:
doh